[flang][OpenMP] Support reduction of allocatable variables

tblah opened the pull request:
Both arrays and trivial scalars are supported. Both cases must use by-ref reductions because both are boxed.

My understanding of the standards are that OpenMP says that this should follow the rules of the intrinsic reduction operators in fortran, and fortran says that unallocated allocatable variables can only be referenced to allocate them or test if they are already allocated. Therefore we do not need a null pointer check in the combiner region.

tblah committed:
[flang][OpenMP] Support reduction of allocatable variables

Both arrays and trivial scalars are supported.

ye-luo commented:
```
program main
  implicit none

  integer:: i
  REAL(8) :: EtsvdW_period
  REAL(8), DIMENSION(:,:), ALLOCATABLE :: FtsvdW_period,HtsvdW_period
  ALLOCATE(HtsvdW_period(3,3))

  !$omp parallel do reduction(+:EtsvdW_period) reduction(+:HtsvdW_period)
  do i=1,10
  enddo

endprogram
```
got ICE with `flang-new -fopenmp main.f90`
```
$ flang-new -fopenmp main.f90
flang-new: /home/yeluo/opt/llvm-clang/llvm-project/llvm/lib/IR/Instructions.cpp:3292: static BinaryOperator *llvm::BinaryOperator::Create(BinaryOps, Value *, Value *, const Twine &, Instruction *): Assertion `S1->getType() == S2->getType() && "Cannot create binary operator with two operands of differing type!"' failed.
```
If I fuse two reductions.
```
!$omp parallel do reduction(+:EtsvdW_period,HtsvdW_period)
```
the code compiles.

ye-luo commented:
OK the issue doesn't seem directly related. reproducible on main https://github.com/llvm/llvm-project/issues/88935

tblah committed:
Correctly handle unallocated allocatables

tblah commented on flang/lib/Lower/OpenMP/ReductionProcessor.cpp:
```diff
@@ -464,9 +501,24 @@ createReductionInitRegion(fir::FirOpBuilder &builder, mlir::Location loc,
+      createReductionCleanupRegion(builder, loc, reductionDecl);
+      builder.restoreInsertionPoint(insPt);
+      return boxAlloca;
+    }
```
Thanks for pointing this out. I think I have now covered these with the addition of https://github.com/llvm/llvm-project/pull/88392/commits/83614c3d2a60324cc32ed04b8b47dc85d1cf3844

tblah commented:
> OK the issue doesn't seem directly related. reproducible on main #88935 After working around the syntax parsing bug, Quantum ESPRESSO compiles although I don't have a test case to hit the actually code path.

Thanks for reporting the issue and trying the patch!

tblah commented:
There is a crash trying to find the right block to place an alloca during FIR lowering after the most recent changes. I will change the MR to WIP until this is fixed.

tblah committed:
Fix broken alloca location

The change that led to this bug being exposed is that we now generate
allocas nested inside of if statements instead of only in the immediate
body of the reduction declare operation.

tblah commented:
The bug should be fixed with https://github.com/llvm/llvm-project/pull/88392/commits/4dc7bd68d9e3e3da8ab2213346f4e24920713485. This is ready for review again

tblah committed:
Merge remote-tracking branch 'origin/main' into ecclescake/allocatable-array-reduction

kiranchandramohan reviewed (approved):
LGTM. Nice work Tom.

There was some effort to centralize code for deallocations.
https://github.com/llvm/llvm-project/pull/67003
There are also functions like `genIsAllocatedOrAssociatedTest` that could be helpful.

tblah commented:
> LGTM. Nice work Tom.
> 
> There was some effort to centralize code for deallocations. #67003 There are also functions like `genIsAllocatedOrAssociatedTest` that could be helpful.

Thanks for the recommendation. Using these APIs would require wrapping everything in a `fir::MutableBoxValue`. I didn't like that approach for a few reasons:
1) I feel it unnecessarily obfuscates what is being done in this case. I already use standard helpers to determine if the pointer is null. E.g. all that `genIsAllocatedOrAssociated` does is get the address and use the same helper to see if it is null.
2) It will make the code structure here more complex because only the allocatable boxes can be a `MutableBoxValue`
3) I want control of the generated fir::IfOp

A lot of the added complexity in the MutableBox code is for running finalizers etc. We only support trivial types here.

kiranchandramohan commented:
> > LGTM. Nice work Tom.
> > There was some effort to centralize code for deallocations. #67003 There are also functions like `genIsAllocatedOrAssociatedTest` that could be helpful.
> 
> Thanks for the recommendation. Using these APIs would require wrapping everything in a `fir::MutableBoxValue`. I didn't like that approach for a few reasons:
> 
> 1. I feel it unnecessarily obfuscates what is being done in this case. I already use standard helpers to determine if the pointer is null. E.g. all that `genIsAllocatedOrAssociated` does is get the address and use the same helper to see if it is null.
> 2. It will make the code structure here more complex because only the allocatable boxes can be a `MutableBoxValue`
> 3. I want control of the generated fir::IfOp
> 
> A lot of the added complexity in the MutableBox code is for running finalizers etc. We only support trivial types here.

OK.

BTW, do we have a TODO for types with finalizers?

tblah commented:
> BTW, do we have a TODO for types with finalizers?

Not a special one. I think it would hit the generic unsupported type in a reduction TODO.
